Wikipedia/View Controllers/WebView/NativeSectionTitleOverlays/WMFSectionTitlesViewController.m

```diff
@@ -0,0 +1,231 @@
+}
+
+- (void)setupTopStaticNativeTitleOverlayLabel {
+    if (self.topStaticNativeTitleLabel) {
+        return;
+    }
+    self.topStaticNativeTitleLabel       = [[WMFTitleOverlayLabel alloc] init];
+    self.topStaticNativeTitleLabel.alpha = 0;
+    [self.webViewController.view addSubview:self.topStaticNativeTitleLabel];
+    [self.topStaticNativeTitleLabel mas_makeConstraints:^(MASConstraintMaker* make) {
+        make.left.equalTo(self.webViewController.view.mas_left);
+        make.right.equalTo(self.webViewController.view.mas_right);
+    }];
+
+    self.topStaticNativeTitleLabelTopConstraint =
+        [NSLayoutConstraint constraintWithItem:self.topStaticNativeTitleLabel
+                                     attribute:NSLayoutAttributeTop
+                                     relatedBy:NSLayoutRelationEqual
+                                        toItem:self.webViewController.topLayoutGuide
+                                     attribute:NSLayoutAttributeBottom
+                                    multiplier:1.0
+                                      constant:0];
+    [self.webViewController.view addConstraint:self.topStaticNativeTitleLabelTopConstraint];
+}
+
+- (void)updateOverlaysPositions {
+    NSArray* headingsTopOffsets = [self getSectionTitlesLocationsJSON];
+    if (headingsTopOffsets.count == self.models.count) {
+        for (NSUInteger i = 0; i < self.models.count; i++) {
+            WMFTitleOverlayModel* m = self.models[i];
+            if (m.anchor && m.topConstraint) {
+                NSNumber* topOffset = headingsTopOffsets[i];
+                CGFloat topFloat    = topOffset.floatValue + self.webView.scrollView.contentOffset.y;
+                m.topConstraint.constant = topFloat;
+                m.yOffset                = topFloat;
+            }
+        }
+    }
+}
+
+- (void)updateTopOverlayForScrollOffsetY:(CGFloat)offsetY {
+    [self updateIndexOfNativeTitleLabelNearestTopForScrollOffsetY:offsetY];
+    [self nudgeTopStaticTitleLabelIfNecessaryForScrollOffsetY:offsetY];
+}
+
+- (void)updateIndexOfNativeTitleLabelNearestTopForScrollOffsetY:(CGFloat)offsetY {
+    self.topStaticNativeTitleLabel.alpha = (offsetY <= 0) ? 0 : 1;
+
+    CGFloat lastOffset          = 0;
+    NSUInteger newTopLabelIndex = -1;
+
+    for (NSUInteger thisIndex = 0; thisIndex < self.models.count; thisIndex++) {
+        WMFTitleOverlayModel* m = self.models[thisIndex];
+        CGFloat thisOffset      = m.yOffset;
+        if (offsetY > lastOffset && offsetY <= thisOffset) {
+            newTopLabelIndex = thisIndex - 1;
+            break;
+        } else if ((thisIndex == (self.models.count - 1)) && (offsetY > thisOffset)) {
+            newTopLabelIndex = thisIndex;
+            break;
+        }
+        lastOffset = thisOffset;
+    }
+
+    if (newTopLabelIndex != -1) {
+        if (newTopLabelIndex != self.indexOfNativeTitleLabelNearestTop) {
+            self.indexOfNativeTitleLabelNearestTop = newTopLabelIndex;
```

Comment:
Instead of setting state and then calling `updateTopStaticTitleLabelText` it would be better to just pass the desired value into the method:

```
- (void) updateTopStaticSectionHeaderWithModel:(WMFTitleOverlayModel*)model;
```

Then you can remove the ivar entirely. 

We should always strive to not have an ivar if possible - any state we have to maintain and update is expensive. if you can just calculate it in place and "forget about it" it makes the code easier to maintain.